### PR TITLE
rls: update picker synchronously on configuration update

### LINKED ISSUE #7212: rls: update picker synchronously upon receipt of configuration update
https://github.com/grpc/grpc-go/issues/5469 recommends an audit of existing LB policies to ensure that they update their pickers synchronously upon receipt of a configuration update.

`rls` LB policy does not update its picker synchronously in one corner case.

- Configuration updates are handled here: https://github.com/grpc/grpc-go/blob/b4f7947184ed45e72283a37aeaf214e38c35bc63/balancer/rls/balancer.go#L252
- The LB policy inhibits picker updates until the new configuration is pushed to all child policies. See: https://github.com/grpc/grpc-go/blob/b4f7947184ed45e72283a37aeaf214e38c35bc63/balancer/rls/balancer.go#L270
- In a blocking fashion, it does the following
  - pushes updates to child policies
  - waits for picker updates from all the child policies
  - sends a new picker
  - See: https://github.com/grpc/grpc-go/blob/b4f7947184ed45e72283a37aeaf214e38c35bc63/balancer/rls/balancer.go#L293
- At the end of `UpdateClientConnState`, the LB policy handles any changes to the data cache size here: https://github.com/grpc/grpc-go/blob/b4f7947184ed45e72283a37aeaf214e38c35bc63/balancer/rls/balancer.go#L297
  - The return value from the resize operation is ignored here: https://github.com/grpc/grpc-go/blob/b4f7947184ed45e72283a37aeaf214e38c35bc63/balancer/rls/balancer.go#L306
  - The return value should not be ignored. Instead if any entry with a backoff timer was evicted, then it should result in a new picker being sent upwards.

Quoting this from the design:
```
When removing an entry whose backoff_time_ is in the future, the backoff timer 
will be cancelled.  Note that just like when the backoff timer fires normally, a new
 picker is returned to the channel, to force it to re-process any wait-for-ready RPCs
that may still be queued if we failed them while we were in backoff.
```

### PR DESCRIPTION
**Pull request by aranjans**
fixes #7212.

### Problem
#5469 recommends an audit of existing LB policies to ensure that they update their pickers synchronously upon receipt of a configuration update. RLS lb policy do updates the picker synchronously but misses out in one single corner case when there is a config change in cache size, where we need to send picker update in case any cache entry with valid backoff was evicted.

### What does this PR do?
This PR handles that corner cases and make sure it updates the picker in all of the cases synchronously with config update of RLS lb policy.

RELEASE NOTES:
- rls: update picker synchronously on configuration update.

### CODE REVIEW

File: balancer/rls/balancer.go
Comment (arjan-bal) [inline balancer/rls/balancer.go:311]: Since we are interested if testing that the picker was updated, we should put the hook inside `sendNewPickerLocked`. Presently, someone can delete the line `b.sendNewPickerLocked()` and the test would still pass as long as the call to `entryWithValidBackoffEvicted()` is there below.

**Review by arjan-bal — COMMENTED** [review]
Great progress since the last attempt!

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:730]: We would need the third backend as well, bcz first one won't be successful RPC as control channel will throw error and put that in backoff. When we make RPC call to backend2 and backend3, it will be actually go into the cache and gets evicted on the config update with resizing of cache.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:664]: thAnks for the catch, you're right. We should check for the picker update before lb config update is complete which would make sure these udpates are synchronous.

**Comment by aranjans** [comment]
@arjan-bal I have updated the PR and referenced your comments in the code change.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:683]: This needs to be a blocking channel if we want to test for synchronous updates, see the comment above.
```go
backOffItemEvicted := make(chan struct{})
```

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:763]: I believe this can be removed as there are only 2 backends now.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:738]: I believe this can be removed now.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:813]: Is this still needed?

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:689]: We should block until the test reads from the channel. This is how we verify that the update doesn't complete before the picker is updated.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:825]: There should be another case for `<-clientConnUpdateDone` which should result in an error that `client conn update was completed before the picker update`.

**Comment by aranjans** [comment]
@arjan-bal thAnks for your review, I have updated the PR with your suggestions. 
fyi: For all the suggestions, I have just put the thumbs up to the comment and updated the PR.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go]: yes,  we'd need this RPC call. I have verified it without having this RPC call, and the test fails.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:812]: Instead of maintaining a counter and and a channel, this can be done using a blocking channel.
```go
pickerUpdateChan := make(chan struct{})
newPickerGenerated = func() {
	newPickerGenerated<-struct{}{}
}

// snipped
for i := 0; i < 2; i++{
	select {
	case <-pckrSentBeforeClientConnUpdate:
	case <-clientConnUpdateDone:
		t.Fatalf("Client conn update was completed before picker update.")
	case <-ctx.Done():
		t.Fatalf("client conn update could not complete: %v", ctx.Err().Error())
	}
}
```

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:803]: Even if we were using a counter, we could move this to the beginning and check for the difference in the counter values before and after calling `UpdateClientConnState`. 

As described in the comment below, I don't think we need a counter. We should move this to the top and verify the order of pickerGeneration above also.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [reply balancer/rls/balancer_test.go]: I was referring to the call with metadata values `"n3": "v3"` which was expected to go to backend 3 before the changes. I see that its removed now.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:825]: We need to wait for `<-clientConnUpdateDone` at the end.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:812]: Actually that might not be correct as we are making the RPC calls in between, so that could also make the picker updated. For ease of the test I think the current state of PR makes more sense to me, as it specifically verifies for status updates that there should be only 2 picker updates happening on cache resizing. WDYT?

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:799]: This should be a blocking channel.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [reply balancer/rls/balancer_test.go:807]: Usually test setup happens before the test is run and results are verified. 
If you want to prevent the blocking till the last update, you can use an atomic boolean like `blockUpdates := atomic.Bool{}`.

In the picker update hook, you can check the bool and let the request through. Just before calling `UpdateState`, you can set the boolean to true.

Comment (arjan-bal) [reply balancer/rls/balancer_test.go:807]: Also reading from the blocking channel twice is a cleaner way than having the hook send one message on the channel after 2 calls.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:824]: Its probably a good idea to wait for context timeout along with this to ensure the test doesn't hang for a long time in case of failures.

**Comment by arjan-bal** [comment]
Please address comments before re-assigning to me :)

**Comment by aranjans** [comment]
Ah my bad, chrome tab wasn't refreshed and i thoughts its still not reviewed.

**Review by arjan-bal — COMMENTED** [review]
For future, please ensure comments made in one place are resolved throughout the PR when applicable. I feel that the similar changes are requested a couple times now. This is dragging on the review for longer than necessary.

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [reply balancer/rls/balancer_test.go]: This wasn't addressed.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:665]: Why do we need to wait for the picker update twice?
We are already blocking updates until the last update using atomic.Bool{}, so we will only have one struct going into `pckrSentBeforeClientConnUpdate`

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:826]: nit: Capital E in "Error waiting.."

File: balancer/rls/balancer_test.go
Comment (arjan-bal) [inline balancer/rls/balancer_test.go:835]: nit: "Client conn..."

**Review by arjan-bal — APPROVED** [review]
LGTM, please address nits and get a second approval before merging.

File: balancer/rls/balancer_test.go
Comment (easwars) [reply balancer/rls/balancer_test.go:835]: Same here. It doesn't provide any value to print `ctx.Err().Error()`. That would simply say something like `context deadline exceeded`. Instead change the error string to be of the form I mentioned in the other comment.

**Review by easwars — CHANGES_REQUESTED** [review]

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:789]: Yeah I had discussion for the same with Arjan [here](https://github.com/grpc/grpc-go/pull/7412#discussion_r1680852679). Let me know if you still have any questions.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:804]: Ummm..... Actually `makeTestRPCAndVerifyError` tries to make an RPC call with no metadata and in the callback defined [here](https://github.com/grpc/grpc-go/pull/7412/files#diff-f9b6e8eebb2ad5ee87acfbac326ff4f3a76bb50d25ff66559192b6e3f148c55bR737), this will throw the error as no metadata will match. And this will cause the control channel to throw the error. And that will cause the item to get into backoff. Please correct me if I am wrong. 

Also when i comment out this RPC call at line 803, the test fails which imo was expected as there won't be any item with valid backoff timer in that case.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:798]: `clientConnUpdateDone` is a blocking channel and `clientConnUpdateHook` will be called twice for init/updating RLS lb config. So we'll have to wait for `clientConnUpdateDone` twice, I've tested this. WDYT?

File: balancer/rls/balancer.go
Comment (easwars) [reply balancer/rls/balancer.go:310]: I don't understand. What does the thumbs up mean? I don't see any change being made here to address my comment.

File: balancer/rls/balancer_test.go
Comment (easwars) [reply balancer/rls/balancer_test.go:789]: What I'm saying is that the same thing can be done without having to spawn a goroutine. Instead of calling `	r.InitialState(resolver.State{ServiceConfig: sc})` before the call to `grpc.Dial`, I would do the following:

- Generate the service config just like how it is being done today, at line 779.
- Then, create a grpc channel. No need of spawning a goroutine.
- Then call `r.UpdateState(resolver.State{ServiceConfig: sc})`
- Then block on a read from the `clientConnUpdateDone`

File: balancer/rls/balancer_test.go
Comment (easwars) [reply balancer/rls/balancer_test.go:804]: Ah yes, you are right. Could you please add a comment for this.

File: balancer/rls/balancer_test.go
Comment (easwars) [reply balancer/rls/balancer_test.go:826]: Again thumbs up without any changes.

File: balancer/rls/balancer_test.go
Comment (easwars) [reply balancer/rls/balancer_test.go:835]: And here.

**Review by easwars — CHANGES_REQUESTED** [review]

**Comment by aranjans** [comment]
@easwars I had not pushed the changes that's why i just ack'ed the comments and did not change the assignees. I just had few questions like [here](https://github.com/grpc/grpc-go/pull/7412#discussion_r1699923239).

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go]: Added now.

File: balancer/rls/balancer.go
Comment (aranjans) [reply balancer/rls/balancer.go]: I had not pushed the changes earlier that's why i just ack'ed the comments and did not change the assignees. I just had few questions like https://github.com/grpc/grpc-go/pull/7412#discussion_r1699923239.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:789]: I tried removing the go routine and instead define the client just before calling r.InitialState(), but that overall call was getting stuck somewhere and it wasn't able to initialize the resolver state. I am debugging this issue, so keeping this thread open. I have resolved all the other threads.

File: balancer/rls/balancer.go
Comment (aranjans) [reply balancer/rls/balancer.go]: Updated

**Comment by aranjans** [comment]
@easwars i have made changes in reference to all your comments except [this](https://github.com/grpc/grpc-go/pull/7412#discussion_r1702387695) one(I tried removing the go routine and instead define the client just before calling r.InitialState(), but that overall call was getting stuck somewhere and it wasn't able to initialize the resolver state. I am debugging this issue). Otherwise i feel this PR is ready to be reviewed.

File: balancer/rls/balancer.go
Comment (easwars) [inline balancer/rls/balancer.go:310]: I think we need to think about this a little more. The fact that a configuration update is leading to multiple picker updates is not ideal. It is a little tricky here because of the order of locks.

Comment (aranjans) [reply balancer/rls/balancer.go]: Hi Easwar, based on our discussions offline I have updated the PR and ensured that we are updating picker only once.

File: balancer/rls/balancer.go
Comment (easwars) [inline balancer/rls/balancer.go:334]: On second thoughts, it seems like we don't even have to acquire `stateMu` here. We are not touching any state guarded by `stateMu` here. We are only pushing a message on the `updateCh`. What do you think?

File: balancer/rls/balancer_test.go
Comment (easwars) [inline balancer/rls/balancer_test.go:760]: Based on my earlier comment, we only need two entries here.

File: balancer/rls/balancer_test.go
Comment (easwars) [inline balancer/rls/balancer_test.go:833]: Why is this required?

File: balancer/rls/balancer_test.go
Comment (easwars) [inline balancer/rls/balancer_test.go:796]: This comment is not accurate.

File: balancer/rls/balancer_test.go
Comment (easwars) [inline balancer/rls/balancer_test.go:663]: We don't need a restartable listener for this test. Is there a reason to use this? Instead, you could simply pass `nil` to: `	rlsServer, rlsReqCh := rlstest.SetupFakeRLSServer(t, nil)`

File: balancer/rls/balancer_test.go
Comment (easwars) [inline balancer/rls/balancer_test.go:723]: There is no need for this backend. We only need two backends.

So, ideally your test should:
- Start two backends
- In `rlsServer.SetResponseCallback`
  - if `k1` == `v1`, return `backend1`
  - if `k2` == `v2`, return `backend2`
  - else, return error "no keys in request metadata"
- Make an RPC with no metadata. This should fail. But this will result in a data cache entry which will go into backoff
- Make an RPC with metadata (`k1=v1`). This should succeed and create a data cache entry. 
- Make an RPC with metadata (`k2=v2`). This should succeed and create a data cache entry. 
- Reduce the size of the data cache to `2`. This should evict the entry in backoff.
- Ensure that a single picker update is sent in this case.

Your test is already doing most of the above, but not all. So, please fix it a little.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:760]: Ah this got here while resolving merge conflicts with master.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:796]: Same this also got here while resolving conflict.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go:663]: Yeah my bad it was removed in earlier commits, this got back while resolving merge conflicts. Updated it now.

File: balancer/rls/balancer_test.go
Comment (aranjans) [reply balancer/rls/balancer_test.go]: Yeah this makes sense, my test was already doing earlier as you see [here](https://github.com/grpc/grpc-go/pull/7412/commits/8166fd4858bf274f2f740e37414f16563c77f9eb#diff-f9b6e8eebb2ad5ee87acfbac326ff4f3a76bb50d25ff66559192b6e3f148c55bL730). But while resolving merge conflicts, third backend got added back, apology for this. I have updated the PR now.

File: balancer/rls/balancer.go
Comment (easwars) [inline balancer/rls/balancer.go:340]: This comment is now also at the top of the `if` statement. So, please delete this.

**Review by easwars — APPROVED** [review]

File: balancer/rls/balancer.go
Comment (aranjans) [reply balancer/rls/balancer.go]: Updated ✅

**Review by easwars — APPROVED** [review]

**Pull request merged**
